### ISSUE TITLE
Custom constraints by injecting them into the Calliope Pyomo model

### ISSUE THREAD

**Issue by timtroendle** [calliope-project/calliope#294]
## Feature description

I need to create my own constraints. Because I prefer not to change Calliope’s base code, I use a non-invasive way: I manually inject Pyomo constraints into the Calliope model. Considering the lack of a plug-in system at the current moment, and considering that building such a system would likely require a lot of resources, I am wondering whether this approach or something similar should be made the official or semi-official way of adding own constraints? For me it works well, and I assume I will continue to use this approach in the future. Making this approach official or semi-official requires mainly documentation work.

This may be a duplicate of #100, feel free to close in that case.

The main code is only a few lines, and it is already mentioned briefly in #100:

```python
model = calliope.Model(path_to_model)
model.run(build_only=True)
pyomo_model = model.backend._backend
pyomo_model.my_constraint = po.Constraint(...)
# further modifications
model = model.rerun() # Works, but rerun has some issues. 
# A slightly more complicated version can avoid them.
```

Of course this requires to understand Pyomo, and it requires to understand the Pyomo model that Calliope builds. The former is obvious and I think many people in the field are able to. The latter is possible by studying the “Mathematical formulation” section of the documentation, or the NetCDF model — as it is very close to the Pyomo model. 

Exposing the Pyomo model is a disadvantage. First, because Pyomo becomes part of the official feature set. If Calliope moves away from Pyomo at one point, this feature will break. Second, because the entire Pyomo model becomes part of the Calliope interface and it needs to be maintained accordingly. Third, the user can break the Pyomo model, for example by removing the set `loc_techs` or anything similarly nasty. To overcome theses issues, the method above could be semi-official: it could be advertised in the documentation as a feasible but not officially supported approach.

Only a few things would be necessary to make this an official or semi-official approach: first, most importantly, it should be documented and advertised. Currently it's difficult to know as a user this actually works. Second, either `rerun` should be fixed or the `run` method changed in a way compatible with this approach. Third, especially if the approach is officially supported, the `Calliope.Model` needs to be slightly updated, so that this is part of the official interface and one would not need to access private members of the class.

## Calliope version

0.6.5.dev

**Comment by brynpickering** [comment]
I like the idea of adding it as a stop gap for those who already know some Pyomo. 

Could the interface look something like ``model.backend.add_constraint(constraint_name, constraint_sets, constraint_rule)``? Then just have the po.Constraint(..) bit done inside a `add_constraint`, since all these backend interface functions are Pyomo-specific anyway?
We can then do a sanity check that the sets over which they want the constraint built actually exist within the backend model, before adding the constraint in.

Another thought: I wonder whether it could be better to split 'run' and 'build' such that the model has a new method 'build' which builds the backend model? Would this be useful beyond Pyomo...?

**Comment by timtroendle** [comment]
Not exposing the actual Pyomo model, but adding a `add_constraint` method sounds like a good idea to me. It makes the approach less powerful, but also less error-prone to the user. (And if you _really_ want to, you can still access the full model through the private member `_backend`.) The only downside: without being able to actually _touch_ the Pyomo model and play around with it, it may be difficult to write a constraint. How do I know which sets exist? How do I know which variables exist? etc

I think having `build()`and `run()` as two separate methods is cleaner than the current approach with the `build_only` parameter -- not a big issue though and would be a breaking change.

**Comment by brynpickering** [comment]
What would then be easy enough to do is to have another interface method to generate a e.g. dictionary of parameters/variables/constraints and the sets over which they are built. Something like this:

```
import pyomo
all_objects = {
    objname: {
        i.name: [j.name for j in i.index_set().set_tuple] if i.name + '_index' == i.index_set().name else [i.index_set().name]
        for i in m._backend_model.component_objects()
        if isinstance(i, getattr(pyomo.core.base, objname))
    }
    for objname in ['Var', 'Param', 'Constraint']    
}
```

**Comment by brynpickering** [comment]
Implementation solved in #312, documentation still needed

**Comment by sjpfenninger** [comment]
Won't address as we will remove this functioanlity in 0.7.0 in favour of new YAML-based custom constraints
